[Ray 2.3 release] `cluster_tune_scale_up_down` release test fails

https://buildkite.com/ray-project/release-tests-branch/builds/1303#0185d157-4ce0-43da-a273-478fd7fbf07e

```
2023-01-20 15:32:36,885	INFO tune.py:782 -- Total run time: 1806.15 seconds (1805.96 seconds for the tuning loop).
--
  | Traceback (most recent call last):
  | File "workloads/tune_scale_up_down.py", line 57, in <module>
  | main()
  | File "workloads/tune_scale_up_down.py", line 53, in main
  | assert node_counter.node_counts[-1] == 1, "Cluster didn't scale down to 1 node."
  | AssertionError: Cluster didn't scale down to 1 node.
```

--- a/python/ray/tune/execution/ray_trial_executor.py
+++ b/python/ray/tune/execution/ray_trial_executor.py
@@ -843,8 +843,8 @@ class RayTrialExecutor:
         """Before step() is called, update the available resources."""
         self._resource_updater.update_avail_resources()
 
-    def on_step_end(self) -> None:
-        self._cleanup_cached_actors()
+    def on_step_end(self, search_ended: bool = False) -> None:
+        self._cleanup_cached_actors(search_ended=search_ended)
         self._do_force_trial_cleanup()
 
     def _count_staged_resources(self):
@@ -854,7 +854,9 @@ class RayTrialExecutor:
             counter[resource_request] += 1
         return counter
 
-    def _cleanup_cached_actors(self, force_all: bool = False):
+    def _cleanup_cached_actors(
+        self, search_ended: bool = False, force_all: bool = False
+    ):
         """Clean up unneeded cached actors.
 
         Ray Tune caches actors for re-use to avoid initialization overhead. This is
@@ -884,8 +886,10 @@ class RayTrialExecutor:
         resources for all cached actors. If we cached more actors than we need, we
         terminate the excess actors and free the resources.
         """
-        if not self._staged_trials and not force_all:
-            # If we don't have any staged trials, keep cached actors
+        if not self._staged_trials and not force_all and not search_ended:
+            # If we don't have any staged trials, keep cached actors,
+            # unless cleanup is forced or no new trials are going to be generated
+            # (if the search ended).
             return
 
         staged_resources = self._count_staged_resources()

--- a/python/ray/tune/execution/trial_runner.py
+++ b/python/ray/tune/execution/trial_runner.py
@@ -967,7 +967,7 @@ class TrialRunner:
         self._reconcile_live_trials()
 
         with warn_if_slow("on_step_end"):
-            self.trial_executor.on_step_end()
+            self.trial_executor.on_step_end(search_ended=self._search_alg.is_finished())
         with warn_if_slow("callbacks.on_step_end"):
             self._callbacks.on_step_end(iteration=self._iteration, trials=self._trials)
 

--- a/python/ray/tune/tests/test_trial_runner_3.py
+++ b/python/ray/tune/tests/test_trial_runner_3.py
@@ -75,7 +75,7 @@ class TrialRunnerTest3(unittest.TestCase):
             cnt = self.pre_step if hasattr(self, "pre_step") else 0
             self.pre_step = cnt + 1
 
-        def on_step_end(self):
+        def on_step_end(self, search_ended: bool = False):
             cnt = self.pre_step if hasattr(self, "post_step") else 0
             self.post_step = 1 + cnt
 

--- a/release/cluster_tests/workloads/tune_scale_up_down.py
+++ b/release/cluster_tests/workloads/tune_scale_up_down.py
@@ -1,3 +1,27 @@
+"""Test cluster up/down scaling behavior.
+
+This test should run on a cluster with autoscaling enabled. It assumes 1-3 nodes
+with 4 CPUs each.
+
+We start a Ray Tune run with 3 trials. Each trial uses 4 CPUs, so fills up a node
+completely. This means we will trigger autoscaling after starting up.
+
+The trial on the head node will run for 30 minutes. This is to make sure that
+we have enough time that the nodes for the other two trials come up, complete
+training, and come down before the first trial finishes.
+
+The other two trials will run once their nodes are up, and take 3 minutes each
+to finish. The three minutes have been chosen to make sure that both trials
+run in parallel for some time, i.e. to avoid that both additional trials run on
+only one node.
+
+We keep track of the number of nodes we observe at any point during the run.
+
+Test owner: krfricke
+
+Acceptance criteria: Should have scaled to 3 nodes at some point during the run.
+Should have scaled down to 1 node at the end.
+"""
 from collections import Counter
 import time
 

